util.go

```diff
@@ -18,13 +21,53 @@ func hash64(x uintptr) uint32 {
 	return uint32(x)
 }
 
-// FNV-1a 32-bit hash
-func fnv32(key string) uint32 {
-	hash := fnv32Offset
-	keyLength := len(key)
-	for i := 0; i < keyLength; i++ {
-		hash ^= uint32(key[i])
-		hash *= fnv32Prime
+//go:noescape
+//go:linkname memhash runtime.memhash
+func memhash(p unsafe.Pointer, h, s uintptr) uintptr
+
+func memhashMap(msg interface{}) uint64 {
+	switch val := msg.(type) {
+	case string:
+		strhv := *(*reflect.StringHeader)(unsafe.Pointer(&val))
+
+		return uint64(memhash(unsafe.Pointer(strhv.Data), 0, uintptr(strhv.Len)))
+	case []byte:
+		strhv := *(*reflect.SliceHeader)(unsafe.Pointer(&val))
+
+		return uint64(memhash(unsafe.Pointer(strhv.Data), 0, uintptr(strhv.Len)))
+	case int:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case uint:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case uintptr:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case int8:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case uint8:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case int16:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case uint16:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case int32:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case uint32:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case int64:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case uint64:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case float64:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case float32:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case bool:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case complex128:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
+	case complex64:
+		return uint64(memhash(unsafe.Pointer(&val), 0, unsafe.Sizeof(val)))
 	}
-	return hash
+
+	return uint64(memhash(unsafe.Pointer(&msg), 0, unsafe.Sizeof(msg)))
```

Comment:
This hash function call doesn't work properly for equal structs. Probably it relies on adresses, not only on memory contents since it was producing different values for equal structs.

Reply:
(I made a new commit and I removed a byte slice which this func should not contain at all. )
Yes, I have intended the added support for built-in concrete types , but user defined structs are not one of them . And it is working perfectly fine with built-in concrete types now. But since we don't have generics in Go yet, most likely an empty interface can confuse people a lot and encourage them for adding structs , slices and pointers to struct (and all the other unsupported types) etc. .  I agree that maybe in the future there can be predeclared type generic at there , and can be more developer friendly in terms of which type parameters this function accepts . Thanks for the review!